### ISSUE TITLE
RTL support

### ISSUE THREAD

**Issue by pikhosh** [sharazam/printpdf#126]
Hi! I wanted to know if this library supports RTL languages or not? In terms of writing and also in terms of layout and alignment.

If not, can you please tell me if there is a solution or if you know a library that has this feature?

**Comment by fschutt** [comment]
RTL - do you mean arabic / hebrew scripts? `printpdf` does not do text layout because that is way too complex for just PDF. Use https://crates.io/crates/azul-text-layout with `inline_text_layout.align_children_horizontal(StyleTextAlignmentHorz::Right);`

You can also try `genpdf` (which is built upon printpdf).

**Comment by pikhosh** [comment]
> RTL - do you mean arabic / hebrew scripts?

Yeah!

> You can also try `genpdf` (which is built upon printpdf).

Thanks! I will try `genpdf`!

**Comment by fschutt** [comment]
@pikhosh RTL support is very difficult, I think I tried implementing it in `azul-text-layout`, but the only Rust crate that can reliably do it is `allsorts`, which is a very low-level font shaper.

**Comment by fschutt** [comment]
allsorts supports Arabic, Hindi, Hebrew, Pinyin and other scripts correctly. But it's very low level, so you have to position the glyphs manually. Doing that in a PDF library is simply out of the scope of this library, even though it would be nice to have.

**Comment by pikhosh** [comment]
> @pikhosh RTL support is very difficult, I think I tried implementing it in `azul-text-layout`, but the only Rust crate that can reliably do it is `allsorts`, which is a very low-level font shaper.

Yeah... it seems difficult! Even among JS libraries, RTL support is rare! (or maybe I just don't know how to use them 😅)

**Comment by fschutt** [comment]
I know I once supported Hebrew in azul-text-layout, I also tried supporting Japanese. But I'm not sure if that is still the case.

**Comment by pikhosh** [comment]
> I know I once supported Hebrew in azul-text-layout, I also tried supporting Japanese. But I'm not sure if that is still the case.

`azul-text-layout` is confusing for me! I've never coded in Rust and `azul-text-layout` seems to be for its own GUI library, which confuses me more about what it does to help create PDFs...

**Comment by fschutt** [comment]
@pikhosh you can use it without a GUI, it just does the text layouting: https://github.com/fschutt/printpdf/issues/39#issuecomment-559118743

Yeah, it's a lot of code to integrate, but allsorts is even more code. In short, I don't know of any "solid" text layouting API that does that.

**Comment by adals** [comment]
for ref [swash](https://github.com/dfrg/swash) and  [demo](https://github.com/dfrg/swash_demo) it could help

**Comment by fschutt** [comment]
In `0.8` there is now a text-shaping API using `allsorts` for glyph layout. Not sure if that will to RTL out-of-the-box but I hope.

**Comment by fschutt** [comment]
RTL support is now working properly, I had to enable some font-shaping features, will be released in the next version this week

**Comment by fschutt** [comment]
Fixed in 0.11, Arabic shapes, connects and right-aligns correctly via from_html (verified visually + pdftotext round-trip during the 0.12 audit, 7dcf565). The remaining font-*selection* bug (explicit families unresolved) was a separate issue, also fixed in 7dcf565.

**Comment by fschutt** [comment]
Correction: RTL core shaping was fixed in `1907861` (0.11.0); the explicit font-family selection residual is `500cbd4`.

**Comment by fschutt** [comment]
Update (history split): the explicit font-family selection fix is now `856d3b2` (RTL core remains `1907861`).
